### PR TITLE
chore(custom-resources): remove stepfunctions as a dependency

### PR DESCRIPTION
**Pull request by nija-at**
### Commit Message
chore(custom-resources): remove stepfunctions as a dependency (#7361)

Swap stepfunction dependency with a custom implementation using
`CfnResource`.

Motivation
Construct libraries such as dynamodb and eks depend on this module to
implement its features. However, at the same time, this module depends
on lambda and stepfunctions construct libraries for its implementation.

This creates dependency cycles.
A simple example - to implement the dynamodb stepfunction task, the
`aws-stepfunctions-tasks` module  needs to depend on `aws-dynamodb`.
However, this creates a cycle such as stepfunctions-tasks → dynamodb →
custom-resources → stepfunctions-tasks.
Another instance is to pre-empt the dependency cycle that will be
introduced by #7224. This will create the cycle - batch → ecs →
route53-targets → cognito → custom-resources → stepfunctions-tasks →
batch.

Conceptually, custom-resources should sit at the bottom of the construct
library stack, taking dependencies on very few foundational CDK
libraries.
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json [hunk 1]
```diff
@@ -1024,7 +1022,6 @@
         }
       },
       "DependsOn": [
-        "comamazonawscdkcustomresourcess3assertproviderwaiterstatemachineRoleDefaultPolicy9882AB39",
```
Comment (eladb) [inline packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json:1027]: Why was this removed?

File: packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  readonly maxAttempts: number;
+
+  readonly backoffRate: number;
+}
+
+/**
+ * A very simple StateMachine construct highly customized to the provider framework.
+ * This is so that this package does not need to depend on aws-stepfunctions module.
+ */
+export class StateMachine extends Construct {
```
Comment (eladb) [inline packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts:21]: To avoid confusion, I would not use a general name such as "state machine". Perhaps something like `WaiterStateMachine` or `WaiterStepFunction`.

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.expected.json [hunk 3]
```diff
@@ -2231,7 +2231,7 @@
               },
               "/",
               {
-                "Ref": "AssetParametersc6510ec9618c99417a04af8cc6ed10b424085b76e94a020e26ae0b5cf3178cf6S3Bucket41E299BC"
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/test/integ.eks-cluster.expected.json:2234]: Do we know why those changed? Can we avoid or at least verify that the actual template is exactly the same?

File: packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    const role = new Role(this, 'Role', {
+      assumedBy: new ServicePrincipal('states.amazonaws.com'),
+    });
+    role.addToPolicy(new PolicyStatement({
+      actions: [ 'lambda:InvokeFunction' ],
+      resources: [ props.isCompleteHandler.functionArn, props.timeoutHandler.functionArn ],
+    }));
+
+    const definition = JSON.stringify({
```
Comment (eladb) [inline packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts:35]: I wonder if it makes more sense to use [Stack.toJsonString](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.Stack.html#to-wbr-json-wbr-stringobj-space) instead so that tokens are properly handled.

File: packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    });
+
+    const resource = new CfnResource(this, 'Resource', {
+      type: 'AWS::StepFunctions::StateMachine',
+      properties: {
+        DefinitionString: definition,
+        RoleArn: role.roleArn,
+      }
+    });
+    resource.addDependsOn(role.node.defaultChild as CfnResource);
```
Comment (eladb) [inline packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts:68]: why not `resource.node.addDependency(role)`? It should take care of things, no?

File: packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts (hunk 4 above)
Comment (nija-at) [reply packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts:35]: Nice! I wasn't aware of this.

File: packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  readonly isCompleteHandler: IFunction;
+
+  readonly timeoutHandler: IFunction;
+
+  readonly interval: Duration;
+
+  readonly maxAttempts: number;
+
+  readonly backoffRate: number;
+}
```
Comment (eladb) [inline packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts:15]: Some comments? (I know it's not a public API but still :-))

**Comment by eladb** [comment]
Minor comment, added "do-not-merge"

File: packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts (hunk 6 above)
Comment (nija-at) [reply packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts]: lol. added.
